### PR TITLE
Enable serialized_reader read specific Page by passing row ranges.

### LINKED ISSUE #1976: Enable serialized_reader read specific Page by passing row ranges.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
# Which issue does this PR close?

Closes #1976.

# Rationale for this change
 Part support #1792 
if we want to use page index get row ranges , first use `SerializedFileReader` get pageIndex info, then use this index get 
 row_ranges like below:
``` rust
        //filter `x < 11`
        let filter =
            |page: &PageIndex<i32>| page.max.as_ref().map(|&x| x < 11).unwrap_or(false);

        let mask = index.indexes.iter().map(filter).collect::<Vec<_>>();

        let row_ranges = compute_row_ranges(&mask, locations, total_rows).unwrap();
```
Finally we can pass the `row_ranges` to new API  to read parquet file(datafusion use this way but without `row_ranges`)
```
fn get_record_reader_by_columns_and_row_ranges(
        &mut self,
        mask: ProjectionMask,
        row_ranges: &RowRanges,
        batch_size: usize,
    ) -> Result<ParquetRecordBatchReader> {
```

# What changes are included in this PR?

One example: if we read  col1, col2  and apply filter get the result we need read `row_ranges[20, 80]`,
_For col1:_ 
  we need all data from page1, page2, page3. 
_For col2:_
 after this PR, we will **filter**  page2 and keep page0, page1
     as for page1: need all data
     as for page0: we need part of its row_range(need row align **TODO**)
```
 * rows   col1   col2   col3
 *      ┌──────┬──────┬──────┐
 *   0  │  p0  │      │      │
 *      ╞══════╡  p0  │  p0  │
 *  20  │ p1(X)│------│------│
 *      ╞══════╪══════╡      │
 *  40  │ p2(X)│      │------│
 *      ╞══════╡ p1(X)╞══════╡
 *  60  │ p3(X)│      │------│
 *      ╞══════╪══════╡      │
 *  80  │  p4  │      │  p1  │
 *      ╞══════╡  p2  │      │
 * 100  │  p5  │      │      │
 *      └──────┴──────┴──────┘
 * 
```

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/array_reader/builder.rs [hunk 1]
```diff
@@ -39,33 +39,46 @@ use crate::data_type::{
...
 
 fn build_reader(
     field: &ParquetField,
     row_groups: &dyn RowGroupCollection,
+    row_groups_filter_offset_index: Option<&Vec<Vec<FilterOffsetIndex>>>,
 ) -> Result<Box<dyn ArrayReader>> {
     match field.field_type {
-        ParquetFieldType::Primitive { .. } => build_primitive_reader(field, row_groups),
+        ParquetFieldType::Primitive { .. } => {
+            build_primitive_reader(field, row_groups, row_groups_filter_offset_index)
```
Comment (Ted-Jiang) [inline parquet/src/arrow/array_reader/builder.rs:75]: For now just support primitive_reader

File: parquet/src/arrow/async_reader.rs [hunk 2]
```diff
@@ -452,7 +454,12 @@ impl RowGroupCollection for InMemoryRowGroup {
         self.row_count
     }
 
-    fn column_chunks(&self, i: usize) -> Result<Box<dyn PageIterator>> {
+    fn column_chunks(
+        &self,
+        i: usize,
+        row_groups_filter_offset_index: Option<&Vec<FilterOffsetIndex>>,
+    ) -> Result<Box<dyn PageIterator>> {
+        //todo support page level filter
```
Comment (Ted-Jiang) [inline parquet/src/arrow/async_reader.rs:462]: will support in InMemoryReader

**Comment by liukun4515** [comment]
I'm confused about the design of the new API described above
```
fn get_record_reader_by_columns_and_row_ranges(
        &mut self,
        mask: ProjectionMask,
        row_ranges: &RowRanges,
        batch_size: usize,
    ) -> Result<ParquetRecordBatchReader> {
```
I think column index reader should be a function for parquet reader or parquet-rs, any one who call the parquet reader should get the benefit from this optimization with a filter.

From your implementation, I find user need to call the lower api and use the column index to calculate the `ranges`.  If so, Any user who want to use the column index of the parquet should add complex custom logic to fit this lower interface.
What is your option? 
@sunchao  @tustvold @viirya

**Comment by sunchao** [comment]
Yes, I think the row ranges are internal to `parquet-rs` and should be calculated during the predicate pushdown.

File: parquet/src/column/reader.rs [hunk 3]
```diff
@@ -160,9 +163,14 @@ where
             num_buffered_values: 0,
             num_decoded_values: 0,
             values_decoder,
+            selected_row_ranges: None,
         }
     }
 
+    pub(crate) fn set_row_ranges(&mut self, row_ranges: RowRanges) {
+        self.selected_row_ranges = Some(row_ranges);
```
Comment (Ted-Jiang) [inline parquet/src/column/reader.rs:171]: Todo need this row_ranges for row align

File: parquet/src/file/serialized_reader.rs [hunk 4]
```diff
@@ -353,6 +392,31 @@ impl<'a, R: 'static + ChunkReader> RowGroupReader for SerializedRowGroupReader<'
         Ok(Box::new(page_reader))
     }
 
+    fn get_column_page_reader_with_offset_index(
```
Comment (Ted-Jiang) [inline parquet/src/file/serialized_reader.rs:395]: Cause of lack test data in sub project in `parquet-testing`, will add end to end test after add test file in it.

Comment (Ted-Jiang) [reply parquet/src/file/serialized_reader.rs:395]: related https://github.com/apache/parquet-testing/pull/25

Comment (liukun4515) [reply parquet/src/file/serialized_reader.rs:395]: We can add UT using the writer API.
https://github.com/apache/arrow-rs/pull/1935 has been merged, the parquet file contains the column index by default.

File: parquet/src/file/page_index/filer_offset_index.rs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    // Note: If data pages are adjacent, will merge them to one `OffsetRange`,
+    // for reduce seek effect.
+    pub(crate) fn calculate_offset_range(&self, row_group_offset: i64) -> OffsetRange {
+        let mut start_list = vec![];
+        let mut length_list = vec![];
+        let page_count = self.get_page_count();
+        if page_count > 0 {
+            //Not after filter page offset
+            let first_page_offset = self.offset_index[0].offset;
+            // add dictionary page if required
```
Comment (liukun4515) [inline parquet/src/file/page_index/filer_offset_index.rs:119]: Why do you assume that the dict page is in the first page of the column chunk?

Comment (liukun4515) [reply parquet/src/file/page_index/filer_offset_index.rs:119]: You can make it clear by change name and type of parameter.
like:
```
calculate_offset_range(&self, directory_page_offset: Option<i64>)
```

File: parquet/src/file/reader.rs [hunk 6]
```diff
@@ -48,6 +49,14 @@ pub trait ChunkReader: Length + Send + Sync {
     /// get a serialy readeable slice of the current reader
     /// This should fail if the slice exceeds the current bounds
     fn get_read(&self, start: u64, length: usize) -> Result<Self::T>;
+
+    /// get a serially readable slice of the current reader
+    /// This should fail if the slice exceeds the current bounds
+    fn get_multi_range_read(
+        &self,
+        start_list: Vec<usize>,
+        length_list: Vec<usize>,
```
Comment (liukun4515) [inline parquet/src/file/reader.rs:58]:
```suggestion
        page_offsets: Vec<usize>,
        page_lengths: Vec<usize>,
```

File: parquet/src/file/reader.rs [hunk 7]
```diff
@@ -86,6 +95,14 @@ pub trait RowGroupReader: Send + Sync {
     /// Get page reader for the `i`th column chunk.
     fn get_column_page_reader(&self, i: usize) -> Result<Box<dyn PageReader>>;
 
+    /// Get page reader for the `i`th column chunk with offset index(in one row group).
+    /// `row_group_pages_offset_index` construct from pageIndex and input rowRanges, here we skip needless page.
+    fn get_column_page_reader_with_offset_index(
+        &self,
+        column_index: usize,
+        row_group_pages_offset_index: &[FilterOffsetIndex],
```
Comment (liukun4515) [inline parquet/src/file/reader.rs:103]:
```suggestion
      page_offset_index : &FilterOffsetIndex,
```
a reference of struct is enough.
I can't figure out a good name for parameter.

File: parquet/src/arrow/array_reader/mod.rs [hunk 8]
```diff
@@ -102,8 +109,16 @@ impl RowGroupCollection for Arc<dyn FileReader> {
         self.metadata().file_metadata().num_rows() as usize
     }
 
-    fn column_chunks(&self, column_index: usize) -> Result<Box<dyn PageIterator>> {
-        let iterator = FilePageIterator::new(column_index, Arc::clone(self))?;
+    fn column_chunks(
+        &self,
+        i: usize,
```
Comment (liukun4515) [inline parquet/src/arrow/array_reader/mod.rs:114]:
```suggestion
        column_index: usize,
```

File: parquet/src/file/serialized_reader.rs [hunk 9]
```diff
@@ -248,21 +286,23 @@ impl<R: 'static + ChunkReader> SerializedFileReader<R> {
...
 
         if options.enable_page_index {
-            //Todo for now test data `data_index_bloom_encoding_stats.parquet` only have one rowgroup
-            //support multi after create multi-RG test data.
-            let cols = metadata.row_group(0);
-            let columns_indexes =
-                index_reader::read_columns_indexes(&chunk_reader, cols.columns())?;
-            let pages_locations =
-                index_reader::read_pages_locations(&chunk_reader, cols.columns())?;
+            let mut columns_indexes = vec![];
```
Comment (liukun4515) [inline parquet/src/file/serialized_reader.rs:289]: Column index should be loaded by lazy.
From the open/read the parquet file, we just load all the column index of the filtered rowgroup.

File: parquet/src/file/serialized_reader.rs [hunk 10]
```diff
@@ -248,21 +286,23 @@ impl<R: 'static + ChunkReader> SerializedFileReader<R> {
...
+            let mut offset_indexes = vec![];
+            for rg in &filtered_row_groups {
+                let c = index_reader::read_columns_indexes(&chunk_reader, rg.columns())?;
+                let p = index_reader::read_pages_locations(&chunk_reader, rg.columns())?;
+
+                columns_indexes.push(c);
+                offset_indexes.push(p);
```
Comment (liukun4515) [inline parquet/src/file/serialized_reader.rs:296]:
```suggestion
                let column_index = index_reader::read_columns_indexes(&chunk_reader, rg.columns())?;
                let offset_index = index_reader::read_pages_locations(&chunk_reader, rg.columns())?;

                columns_indexes.push(c);
                offset_indexes.push(p);
```

File: parquet/src/file/serialized_reader.rs [hunk 11]
```diff
@@ -248,21 +286,23 @@ impl<R: 'static + ChunkReader> SerializedFileReader<R> {
...
-            //support multi after create multi-RG test data.
-            let cols = metadata.row_group(0);
-            let columns_indexes =
-                index_reader::read_columns_indexes(&chunk_reader, cols.columns())?;
-            let pages_locations =
-                index_reader::read_pages_locations(&chunk_reader, cols.columns())?;
+            let mut columns_indexes = vec![];
+            let mut offset_indexes = vec![];
+            for rg in &filtered_row_groups {
+                let c = index_reader::read_columns_indexes(&chunk_reader, rg.columns())?;
```
Comment (liukun4515) [inline parquet/src/file/serialized_reader.rs:292]: If a schema has `co1,col2,col3.....col8`, and we just need the `col1` and `col3`, do we need to load other useless index data？

File: parquet/src/arrow/arrow_reader.rs [hunk 12]
```diff
@@ -88,10 +98,15 @@ impl ArrowReaderOptions {
     ///
 
     /// Set `skip_arrow_metadata` to true, to skip decoding this
-    pub fn with_skip_arrow_metadata(self, skip_arrow_metadata: bool) -> Self {
-        Self {
-            skip_arrow_metadata,
-        }
+    pub fn with_skip_arrow_metadata(mut self, skip_arrow_metadata: bool) -> Self {
+        self.skip_arrow_metadata = skip_arrow_metadata;
+        self
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader.rs:103]:
```suggestion
        {
            skip_arrow_metadata,
            ..self
        }
```

And same below

File: parquet/src/arrow/arrow_reader.rs [hunk 13]
```diff
@@ -130,17 +145,73 @@ impl ArrowReader for ParquetFileArrowReader {
...
-            self.file_reader
-                .metadata()
-                .file_metadata()
-                .schema_descr_ptr(),
-            Arc::new(self.get_schema()?),
-            mask,
-            Box::new(self.file_reader.clone()),
-        )?;
+        if self.options.selected_rows.is_some() {
+            let ranges = &self.options.selected_rows.as_ref().unwrap().clone();
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader.rs:149]:
```suggestion
        if let Some(ranges) = self.options.selected_rows.as_ref()
```

File: parquet/src/arrow/arrow_reader.rs [hunk 14]
```diff
@@ -68,11 +70,19 @@ pub trait ArrowReader {
         mask: ProjectionMask,
         batch_size: usize,
     ) -> Result<Self::RecordReader>;
+
+    fn get_record_reader_by_columns_and_row_ranges(
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader.rs:74]: Do we need this, or is the ArrowReaderOptions sufficient?

File: parquet/src/file/metadata.rs [hunk 15]
```diff
@@ -55,8 +55,8 @@ use crate::schema::types::{
 pub struct ParquetMetaData {
     file_metadata: FileMetaData,
     row_groups: Vec<RowGroupMetaData>,
-    page_indexes: Option<Vec<Index>>,
-    offset_indexes: Option<Vec<Vec<PageLocation>>>,
+    page_indexes: Option<Vec<Vec<Index>>>,
```
Comment (tustvold) [inline parquet/src/file/metadata.rs:58]:
```suggestion
    /// Page index for all pages in each column chunk
    page_indexes: Option<Vec<Vec<Index>>>,
```
Or something like that, same for the below

File: parquet/src/file/reader.rs [hunk 16]
```diff
@@ -48,6 +49,14 @@ pub trait ChunkReader: Length + Send + Sync {
     /// get a serialy readeable slice of the current reader
     /// This should fail if the slice exceeds the current bounds
     fn get_read(&self, start: u64, length: usize) -> Result<Self::T>;
+
+    /// get a serially readable slice of the current reader
+    /// This should fail if the slice exceeds the current bounds
+    fn get_multi_range_read(
```
Comment (tustvold) [inline parquet/src/file/reader.rs:55]: As discussed on #1955  I'm not a fan of this, I would much rather the page reader reads pages, than skipping byte ranges behind its back.

It also changes the semantics of how a column chunk is read, as it now buffers in memory an extra time

File: parquet/src/file/serialized_reader.rs [hunk 17]
```diff
@@ -82,6 +91,27 @@ impl ChunkReader for Bytes {
...
+                "Actual start_list size doesn't match the length_list size ({} vs {})",
+                start_list.len(),
+                length_list.len()
+            ));
+        } else {
+            let mut combine_vec: Vec<u8> = vec![];
+            for (start, length) in start_list.into_iter().zip(length_list.into_iter()) {
+                combine_vec.extend(self.slice(start..start + length).to_vec());
+            }
+            let reader = Bytes::copy_from_slice(combine_vec.as_slice()).reader();
```
Comment (tustvold) [inline parquet/src/file/serialized_reader.rs:111]: This adds an additional copy of all the page bytes, which is definitely not ideal...

File: parquet/src/file/page_index/filer_offset_index.rs [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+use crate::file::page_index::range::{Range, RowRanges};
+use parquet_format::PageLocation;
+
+/// Returns the filtered offset index containing only the pages which are overlapping with rowRanges.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct FilterOffsetIndex {
+    // read from parquet file which before the footer.
+    offset_index: Vec<PageLocation>,
+
+    // use to keep needed page index.
```
Comment (tustvold) [inline parquet/src/file/page_index/filer_offset_index.rs:27]: Why is this necessary?

**Review by tustvold — COMMENTED** [review]
I've had a quick review, unfortunately I think this is missing a key detail. In particular the arrow writer must read the same records from each of its columns. As written this simply skips reading pruned pages from columns. There is no relationship between the page boundaries across columns within a parquet, and therefore this will return different rows for each of the columns.

As described in https://github.com/apache/arrow-rs/pull/1791#pullrequestreview-996352857, you will need to extract the row selection in addition to the page selection, and push this into RecordReader and ColumnValueDecoder. This will also make the API clearer, as we aren't going behind their back and skipping pages.

**Comment by Ted-Jiang** [comment]
> I've had a quick review, unfortunately I think this is missing a key detail. In particular the arrow writer must read the same records from each of its columns. As written this simply skips reading pruned pages from columns. There is no relationship between the page boundaries across columns within a parquet, and therefore this will return different rows for each of the columns.

Thanks @tustvold, your are right. Maybe I made the title confusing😭. as you mentioned in  [#1791 (review)]. (https://github.com/apache/arrow-rs/pull/1791#pullrequestreview-996352857):

>Pass row selection down to RecordReader
>Add a skip_next_page to PageReader
>Add a skip_values to ColumnValueDecoder

This pr is only about the `skip_next_page` part, we will only return the needed page metadata in iterator. As make the  same records from each of its columns (row align), i prefer support in next pr. I prefer to separate them to avoid huge PR and conflict. If you prefer to combine them, I will make this in progress and keep developing.

> As described in [#1791 (review)](https://github.com/apache/arrow-rs/pull/1791#pullrequestreview-996352857), you will need to extract the row selection in addition to the page selection, and push this into RecordReader and ColumnValueDecoder. This will also make the API clearer, as we aren't going behind their back and skipping pages at the block-level
As above, need pass the `row_ranges` to ColumnValueReader in future.

File: parquet/src/file/page_index/filer_offset_index.rs (hunk 18 above)
Comment (Ted-Jiang) [reply parquet/src/file/page_index/filer_offset_index.rs:27]: We pass all `PageLocation` and `RowRanges` in this struct then do the filter logic.

if we have 5 pages, in `try_new`, we filter 2 pages and keep these 3 pages index_numbers in this `index_map` for final `calculate_offset_range`.

**Comment by tustvold** [comment]
I think small incremental PRs is a good approach. However, I have concerns with this specific PR:

* It introduces public APIs that don't have clear semantics (the skipped rows are somewhat arbitrary)
* I would prefer an approach that collocates the page and row skipping logic, instead of treating them as separate concerns. Once RecordReader is skipping rows it will be incredibly confusing if pages are being skipped somewhere else in addition

I wonder if a plan of attack akin to the following might work:

* Add a skip page function to SerializedPageReader that uses the column index to skip the next page without reading it (we may need to change it to take ChunkReader instead of Read)
* Same as the above for InMemoryPageReader
* Add the ability to skip decoding rows to ColumnValueDecoder
* Pass index and row selection down to RecordReader
* Perform skipping

Currently it feels like we're adding the high-level functionality before the necessary lower level functionality exists...

**Comment by tustvold** [comment]
> I got worried about where should i pass the column index info

Give me a day or so and I'll get a PR up with some stuff stubbed out, I think this exercise will help us both :smile:

**Comment by tustvold** [comment]
Marking as a draft, as I think the approach in #1998 is what we will take forward

**Comment by tustvold** [comment]
@Ted-Jiang Can this be closed now?
